src/server/frontend_wayland/wlr_screencopy_v1.cpp

```diff
@@ -0,0 +1,264 @@
+#include "wayland_timespec.h"
+#include "output_manager.h"
+
+#include <boost/throw_exception.hpp>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace mg = mir::graphics;
+namespace geom = mir::geometry;
+
+namespace mir
+{
+namespace frontend
+{
+struct WlrScreencopyV1Ctx
+{
+    std::shared_ptr<Executor> const wayland_executor;
+    OutputManager& output_manager;
+    std::shared_ptr<graphics::GraphicBufferAllocator> const allocator;
+    std::shared_ptr<compositor::ScreenShooter> const screen_shooter;
+
+    auto extents_of_output(wl_resource* output) -> geom::Rectangle;
+};
+
+class WlrScreencopyManagerV1Global
+    : public wayland::WlrScreencopyManagerV1::Global
+{
+public:
+    WlrScreencopyManagerV1Global(wl_display* display, std::shared_ptr<WlrScreencopyV1Ctx> const& ctx);
+
+private:
+    void bind(wl_resource* new_resource) override;
+
+    std::shared_ptr<WlrScreencopyV1Ctx> const ctx;
+};
+
+class WlrScreencopyManagerV1
+    : public wayland::WlrScreencopyManagerV1
+{
+public:
+    WlrScreencopyManagerV1(wl_resource* resource, std::shared_ptr<WlrScreencopyV1Ctx> const& ctx);
+
+private:
+    void capture_output(struct wl_resource* frame, int32_t overlay_cursor, struct wl_resource* output) override;
+    void capture_output_region(
+        struct wl_resource* frame,
+        int32_t overlay_cursor,
+        struct wl_resource* output,
+        int32_t x, int32_t y,
+        int32_t width, int32_t height) override;
+
+    std::shared_ptr<WlrScreencopyV1Ctx> const ctx;
+};
+
+class WlrScreencopyFrameV1
+    : public wayland::WlrScreencopyFrameV1
+{
+public:
+    WlrScreencopyFrameV1(
+        wl_resource* resource,
+        std::shared_ptr<WlrScreencopyV1Ctx> const& ctx,
+        geometry::Rectangle const& area);
+
+private:
+    /// From wayland::WlrScreencopyFrameV1
+    /// @{
+    void copy(struct wl_resource* buffer) override;
+    void copy_with_damage(struct wl_resource* buffer) override;
+    /// @}
+
+    std::shared_ptr<WlrScreencopyV1Ctx> const ctx;
+    geometry::Rectangle const area;
+    geometry::Stride const stride;
+};
+}
+}
+
+auto mf::create_wlr_screencopy_manager_unstable_v1(
+    wl_display* display,
+    std::shared_ptr<Executor> const& wayland_executor,
+    OutputManager& output_manager,
+    std::shared_ptr<graphics::GraphicBufferAllocator> const& allocator,
+    std::shared_ptr<compositor::ScreenShooter> const& screen_shooter)
+-> std::shared_ptr<wayland::WlrScreencopyManagerV1::Global>
+{
+    auto ctx = std::shared_ptr<WlrScreencopyV1Ctx>{new WlrScreencopyV1Ctx{
+        wayland_executor,
+        output_manager,
+        allocator,
+        screen_shooter}};
+    return std::make_shared<WlrScreencopyManagerV1Global>(display, std::move(ctx));
+}
+
+auto mf::WlrScreencopyV1Ctx::extents_of_output(wl_resource* output) -> geom::Rectangle
+{
+    geom::Rectangle extents;
+    if (!output_manager.with_config(output, [&](mg::DisplayConfigurationOutput const& config)
```

Comment:
Not blocking but... `with_config()` seems a bit terse.

Maybe `with_config_for(output).do([&](auto const& ...`

Reply:
Changed to `with_config_for(output, [&](...`